app/utils/server/lib/normalizeMessagesForUser.js

```diff
@@ -0,0 +1,64 @@
+import { Users } from '../../../models/server';
+import { settings } from '../../../settings/server';
+
+const filterStarred = (message, uid) => {
+	// only return starred field if user has it starred
+	if (message.starred && Array.isArray(message.starred)) {
+		message.starred = message.starred.filter((star) => star._id === uid);
+	}
+	return message;
+};
+
+// TODO: we should let clients get names on demand instead of doing this
+export const normalizeMessagesForUser = (messages, uid) => {
+	// if not using real names, there is nothing else to do
+	if (!settings.get('UI_Use_Real_Name')) {
+		return messages.map((message) => filterStarred(message, uid));
+	}
+
+	const usernames = new Set();
+
+	messages.forEach((message) => {
+		message = filterStarred(message, uid);
+
+		usernames.add(message.u.username);
+
+		if (message.mentions && message.mentions.length) {
+			message.mentions.forEach((mention) => { usernames.add(mention.username); });
+		}
+
+		if (message.reactions && Object.keys(message.reactions).length) {
```

Comment:
Can we use `Object.values(message.reactions)` and cache it to a variable to not use it 2 times and do not need to access via `message.reactions[reaction]`? The same below?

Reply:
hardly we gonna have so many reactions to concern about that, but even concerned...we dont need to test and cache... my suggestion, something more simple like `||`:
```javascript

Object.values(message.reactions ||[])
.forEach((reaction) => reaction.usernames.forEach(usernames.add));